src/main/java/com/laytonsmith/core/compiler/signature/FunctionSignature.java

```diff
@@ -0,0 +1,175 @@
+package com.laytonsmith.core.compiler.signature;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Stack;
+
+import com.laytonsmith.PureUtilities.Common.StringUtils;
+import com.laytonsmith.core.constructs.CClassType;
+import com.laytonsmith.core.constructs.InstanceofUtil;
+import com.laytonsmith.core.environments.Environment;
+import com.laytonsmith.core.natives.interfaces.Mixed;
+
+/**
+ * Represents a single function, procedure or closure signature.
+ * @author P.J.S. Kools
+ */
+public class FunctionSignature {
+
+	private final ReturnType returnType;
+	private final List<Param> params;
+	private final List<Throws> throwsList;
+
+	/**
+	 * Creates a new {@link FunctionSignature} with the given properties.
+	 * @param returnType - The function return type.
+	 * @param params - The function parameters.
+	 * @param throwsList - The list of possibly thrown exceptions by the function.
+	 */
+	public FunctionSignature(ReturnType returnType, List<Param> params, List<Throws> throwsList) {
+		this.returnType = returnType;
+		this.params = params;
+		this.throwsList = throwsList;
+	}
+
+	/**
+	 * Creates a new {@link FunctionSignature} with the given return type and no parameters
+	 * and possibly thrown exceptions.
+	 * @param returnType - The function return type.
+	 */
+	public FunctionSignature(ReturnType returnType) {
+		this(returnType, new ArrayList<>(), new ArrayList<>());
+	}
+
+	protected void addParam(Param param) {
+		this.params.add(param);
+	}
+
+	protected void addThrows(Throws throwsObj) {
+		this.throwsList.add(throwsObj);
+	}
+
+	/**
+	 * Gets the function's return type.
+	 * @return The return type.
+	 */
+	public ReturnType getReturnType() {
+		return this.returnType;
+	}
+
+	/**
+	 * Gets the function's parameters.
+	 * @return The parameters.
+	 */
+	public List<Param> getParams() {
+		return Collections.unmodifiableList(this.params);
+	}
+
+	/**
+	 * Gets the function's possibly thrown exceptions.
+	 * @return The possibly thrown exceptions.
+	 */
+	public List<Throws> getThrows() {
+		return Collections.unmodifiableList(this.throwsList);
+	}
+
+	/**
+	 * Matches the given argument types against this signature.
+	 * @param argTypes - The argument types.
+	 * @param env - The environment.
+	 * @param allowUnmatchedArgs - Whether or not to allow more arguments than the signature prescribes.
+	 * This should be {@code false} for normal functions, and {@code true} for procedures and closures.
```

Comment:
Let's remove this parameter, and treat it as if it's always false in the function. The way it's going to work in MethodScript is that generics can have variadic types. Read https://github.com/EngineHub/CommandHelper/blob/genericsTake2/src/main/resources/docs/Generics#L506 for more information, but in general, expect that closure's type will be defined as `class closure<Return, Params...>` and thus a `new closure<string, int, int>` will only be able to accept 2 arguments. The default for params is `auto...` and so thus is without further specification, allowed to send any number of parameters of any type, but that is still a strictly defined value.

Reply:
Sounds good, I've added it because it is currently supported, and while I believe that `@arguments` is a weird hack for which varargs or an array argument should be used instead, it is a breaking change if we remove it. We could think about adding a setting in file options whether or not to allow passing too many or too few arguments to procs and closures.